Add community applications (trays) fetch support

jmaupetit opened the pull request:
## Purpose

Now that we are able to work with tray-packaged apps, the fun part begin! We now need to be able to download, configure and add community apps that are not part of official apps.

## Proposal

- [x] add `community_apps` target directory
- [x] add `fetch_dependencies.yml` playbook and arnold `install` command

jmaupetit committed:
🔧(apps) add community_apps directory

First step to the implementation of a third party apps installation
feature: the target directory where we will install applications.

jmaupetit committed:
✨(apps) add tray fetch support

Now that we are able to work with tray-packaged apps, the fun part
begin! The current work allows to download and install trays that can be
deployed with Arnold. Yay!

jmaupetit commented on group_vars/all/apps.yml:
```diff
@@ -0,0 +1,22 @@
+#
```
I should annotate/comment this file.

lunika replied:
I'm not sure this file should be versionned. Maybe I'm wrong but I understand that I will have to edit it if I want to add a third party application. If I modify it, it will appear in the list of modified files in my git status and I don;t think we want this.

jmaupetit replied:
You can safely override default variables for a customer/env_type with your own apps repository.

jmaupetit commented on group_vars/all/apps.yml:
```diff
@@ -0,0 +1,22 @@
+  - "community_apps/"
+
+apps_repository:
+  - name: funmooc
```
Should comment this as a template

jmaupetit commented on group_vars/all/apps.yml:
```diff
@@ -0,0 +1,22 @@
+  - "apps/"
+  - "community_apps/"
+
+apps_repository:
```
```suggestion
apps_registry:
```

jmaupetit commented on group_vars/all/apps.yml:
```diff
@@ -0,0 +1,22 @@
+      # Should be https or ssh (requires more configuration) git repository url
+      repository: https://github.com/openfun/fun-mooc.git
+      # Can be a tag, a branch or a commit sha1
+      version: add-arnold-package
```
This should be guessed from the application version listed in `apps` (when #308 will be implemented).

jmaupetit commented on group_vars/all/apps.yml:
```diff
@@ -0,0 +1,22 @@
+core_apps_paths:
+  - "core_apps/"
+
+community_apps_dir: "community_apps"
```
Will be removed once all apps moved from Arnold to their own repository.

lunika commented on tasks/fetch_app.yml:
```diff
@@ -0,0 +1,48 @@
+- name: Install fetched tray
+  copy:
+    src: "{{ build_dir.path }}/{{ tray_finder.tray.path }}/"
+    dest: "{{ community_apps_dir }}/{{ tray.metadata.name }}"
```
You can maybe also removed the temporary clone ?

jmaupetit replied:
Not sure, that is the OS responsibility to delete temporary files (in `/tmp`) not our.

lunika commented on fetch_apps.yml:
```diff
@@ -0,0 +1,15 @@
+    - name: Display playbook name
+      debug: msg="==== Starting fetch_apps playbook ===="
+
+    - import_tasks: tasks/set_vars.yml
```
Is it really needed ?

jmaupetit replied:
I have to test it but probably not.

lunika commented on bin/arnold:
```diff
@@ -194,6 +200,7 @@ COMMANDS:
   create_db_vault   create or update a vault file with databases credentials
   htpasswd          initialize a namespace's htpassword files
   info              display current configuration
+  install           fetch dependencies (deployable community apps)
```
`install` is not really intuitive. I expect something else then fetching dependencies.

Could be `dependencies` ? `dependency` ?

jmaupetit replied:
Would `install_trays` be more explicit?

lunika replied:
yes

lunika reviewed (commented):
We will have to manually maintain a registry in this repo ?

jmaupetit commented:
> We will have to manually maintain a registry in this repo ?

Yup. Until we come up with a better idea.

lunika commented on fetch_dependencies.yml:
```diff
@@ -0,0 +1,63 @@
+      include_vars:
+        file: "{{ apps_registry_file_name }}"
+        name: registry
+      when: registry_file.stat.exists
```
You should also abort if the registry file is not created.

lunika commented on fetch_dependencies.yml:
```diff
@@ -0,0 +1,63 @@
+
+    - name: Load dependencies
+      include_vars:
+        file: "dependencies.yml"
```
can we variabilize this file name ?

jmaupetit replied:
Sure we can, but it's more like a convention.

lunika commented on registry.yml:
```diff
@@ -0,0 +1,20 @@
+# Community apps registry
```
Did you plan to check if this file is correctly formatted ? A doc somewhere with properties we can use ? `subpath` is not always used as I see.

jmaupetit replied:
Good points. I should definitely write a documentation for this. What do you have in mind for its validation? A schema?

lunika replied:
Yes like a schema and validated in the CI

jmaupetit replied:
I've added a JSONSchema specification for the registry file format, and it is now automatically checked by the CI.

lunika commented on tasks/fetch_app.yml:
```diff
@@ -0,0 +1,102 @@
+
+- name: Confirm tray update
+  pause:
+    prompt: "{{ tray.metadata.name }} tray already exists. Do you want to update it? Press enter to continue, Ctrl+C to interrupt"
```
`Ctrl+c` will interrupt all the command or just this app installation ? If I install 3 apps, the first is already installed and I don't want to update it, is it possible to skip it and continue ?

jmaupetit replied:
Unfortunately it will break the loop, so interrupt all the command.

lunika replied:
I think we should not. Is there another way to do it in multiple step ? Asking a value and depending on this value skip remaining tasks ?

jmaupetit replied:
I've implemented this behavior. Thanks for the suggestion. :muscle:

lunika reviewed (changes requested):
Sorry, I don't want to block the PR but I can't undo my previous approval and it was a mistake.

jmaupetit commented:
@lunika I've addressed all your comments. Ready for a last review? :pray:

lunika reviewed (approved):
![](https://media.giphy.com/media/vmtxnxveVUodG/giphy.gif)

jmaupetit committed:
🔖(minor) bump release to 6.5.0

Added:

- Community applications (_i.e._ trays not distributed with Arnold) can now be
  fetched, installed, configured & deployed.
